### ISSUE TITLE
Eradicate all places recast modifies node fields output by the parser, use custom flags like ignoreLoc instead

### ISSUE THREAD

**Issue by jedwards1211** [benjamn/recast#905]
Along the lines of #557, #561, there are a bunch of other places where the code sets a node's `loc` to `null` that haven't been reported in issues or PRs.

I agree with @coderaiser, this a really bad and never should have been done. It just caught me by surprise as well.

```js
    // Nullify .loc information for the child declaration so that we never
    // try to reprint it without also reprinting the export declaration.
    node.declaration.loc = null;
```

```js
    // If the node is a MethodDefinition or a .method or .shorthand
    // Property, then the location information stored in
    // node.value.loc is very likely untrustworthy (just the {body}
    // part of a method, or nothing in the case of shorthand
    // properties), so we null out that information to prevent
    // accidental reuse of bogus source code during reprinting.
    node.value.loc = null;
```

etc.

There are other very questionable AST modifications:

```js
    if (n.FunctionExpression.check(node.value)) {
      // FunctionExpression method values should be anonymous,
      // because their .id fields are ignored anyway.
      node.value.id = null;
    }
```

The fields output by the parser are sacred and recast should not mess with them...it should store any necessary tweaked information in its own custom fields.

In the case of setting `loc` to null, recast is basically just trying to ignore the location with reprinting, so we should just set a flag on the node instead, like `ignoreLoc`, or `extra.ignoreLoc`.

**Comment by benjamn** [comment]
> In the case of setting `loc` to null, recast is basically just trying to ignore the location with reprinting, so we should just set a flag on the node instead, like `ignoreLoc`, or `extra.ignoreLoc`.

@jedwards1211 I agree with this alternative in principle, but I think I lean towards setting `node.loc.hideFromRecast = true` to signal (to `recast.print`) not to use that `node.loc` object. Any problems with that idea that you can see?

**Comment by jedwards1211** [comment]
@jedwards1211 wouldn't cause any problems for me.  Theoretically someone could be using a deep-equal assertion on `node.loc` but I assume it's unlikely.

I wasn't sure if the `FunctionExpression` case I referenced has to do with output formatting, but it made me wonder if other kinds of flags would be necessary for some cases too.  In which case I would think grouping recast's flags together would make more sense

**Comment by jedwards1211** [comment]
@benjamn anyway, once we come to a conclusion about how to do it I could start working on a PR

**Comment by jedwards1211** [comment]
Thinking about it more I just don't like the idea of touching the original node fields at all, even `extra`, since touching the original fields has already caused nasty surprises, and a more principled separation of concerns is easily possible.  Would `node.loc.hideFromRecast` actually fulfill any concrete objectives, or is it just a vague impulse?

To me putting `ignoreLoc` and other recast-specific information into a `recast` property has more concrete benefits:
* if you need to pass the AST to other tools for whatever reason, it's more likely to work.  At worst you would just have to stash the `recast` properties in a map, delete them from the AST nodes, call the other tool, then put them back.
* users would be less confused about what's original and what recast injected

**Comment by benjamn** [comment]
@jedwards1211 My thinking behind `node.loc.hideFromRecast`:
1. It has "Recast" in the name, so it won't be mistaken for something that came from the parser, like `node.loc.ignore` or `node.ignoreLoc` could be.
2. It lives within the `node.loc` object, which seems appropriate because it says something about that object and its suitability for later use.
3. Unless we are careful, Recast may be confused by extra `node` properties like `node.ignoreLoc`, and conclude that the AST node has changed and needs to be reprinted. If the metadata lives in `node.loc` instead, that won't be a problem.
4. We can use the [`TreeCopier`](https://github.com/benjamn/recast/blob/0a1260c2113ef2caf0e0e97c1ec17d7e48726b60/lib/parser.ts#L196) logic to add metadata to every `node.loc` without modifying the original AST (warning: `TreeCopier` is some _old_ pre-TypeScript code). We already add some other properties like `node.loc.lines`, for comparison.

In general, if I understand the gist of your comments, I'm open to grouping all these properties under a `node.loc.recast` object, like `node.loc.recast.lines` and `node.loc.recast.hide`, etc.
